packages/pool/contracts/GetterUtils.sol

```diff
@@ -113,24 +156,28 @@ contract GetterUtils is StateUtils, IGetterUtils {
         override
         returns(uint256)
     {
-        return userShares(userAddress) * totalStake / totalSupply();
+        return userShares(userAddress) * totalStake / totalShares();
     }
 
     /// @notice Called to get the voting power delegated to a user at a
     /// specific block
-    /// @param fromBlock Block number for which the query is being made for
     /// @param userAddress User address
+    /// @param _block Block number for which the query is being made for
     /// @return Voting power delegated to the user at the block
     function userReceivedDelegationAt(
-        uint256 fromBlock,
-        address userAddress
+        address userAddress,
+        uint256 _block
         )
         public
         view
         override
         returns(uint256)
     {
-        return getValueAt(users[userAddress].delegatedTo, fromBlock);
+        Checkpoint[] storage delegatedTo = users[userAddress].delegatedTo;
+        uint256 minimumCheckpointIndex = delegatedTo.length > MAX_INTERACTION_FREQUENCY
```

Comment:
What does this check supposed to do? delegaredTo.length is amount of users this user delegated at some point of time as I understand...

Reply:
`delegatedTo` keeps the total voting power that has been delegated to to that user. The problem with that is that I can repeatedly delegate 1 Wei worth of voting power to a user to bloat their `delegateTo` and when they want to cast a vote, their `balanceOfAt()` will take a very long time to find how much voting power has been delegated to them (this is largely mitigated as an attack surface with https://github.com/api3dao/api3-dao/pull/143 because my repeated delegations will be overwritten, but still). So we want the user's `balanceOfAt()` to only search through the last `MAX_INTERACTION_FREQUENCY` elements in their `delegateTo` and it will be able to find the value that it's looking for there because we are also limiting the additions to that array by `MAX_INTERACTION_FREQUENCY` in an `EPOCH_LENGTH` (here https://github.com/api3dao/api3-dao/blob/optimize-withdrawal/packages/pool/contracts/StateUtils.sol#L484-L487)